src/main/java/org/zalando/nakadi/partitioning/HashPartitionStrategy.java

```diff
@@ -30,17 +49,33 @@ public String calculatePartition(final EventType eventType, final JSONObject eve
 
             final JsonPathAccess traversableJsonEvent = new JsonPathAccess(event);
 
+            final ArrayList<String> fieldValues = new ArrayList<>();
             final int hashValue = partitionKeyFields.stream()
                     // The problem is that JSONObject doesn't override hashCode(). Therefore convert it to
                     // a string first and then use hashCode()
                     .map(pkf -> EventCategory.DATA.equals(eventType.getCategory()) ? DATA_PATH_PREFIX + pkf : pkf)
-                    .map(Try.wrap(okf -> traversableJsonEvent.get(okf).toString().hashCode()))
+                    .map(Try.wrap(okf -> {
+                        final String fieldValue = traversableJsonEvent.get(okf).toString();
+                        fieldValues.add(fieldValue);
+                        return stringHash.hashCode(fieldValue);
+                    }))
                     .map(Try::getOrThrow)
                     .mapToInt(hc -> hc)
                     .sum();
 
-            final int partitionIndex = abs(hashValue) % partitions.size();
-            return partitions.get(partitionIndex);
+
+            int partitionIndex = abs(hashValue) % partitions.size();
+            partitionIndex = hashPartitioningCrutch.adjustPartitionIndex(partitionIndex, partitions.size());
+
+            final List<String> sortedPartitions = partitions.stream().sorted().collect(Collectors.toList());
```

Comment:
@adyach `Collections.sort` sorts the initial collection, it is unmodifiable in our case so it will fail.

@rcillo I decided to use indexes instead of actual partitions not to be bind to Kafka specific stuff.